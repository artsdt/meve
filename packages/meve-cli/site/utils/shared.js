import Vue from 'vue'

export const isArray = val => Array.isArray(val)

export function getMenuName(path) {
  return path.slice(path.lastIndexOf('/') + 1)
}

export function getLang(path) {
  return path.slice(1, path.lastIndexOf('/'))
}

export function buildTreeRelation(tree, parent, seen) {
  const flatTree = []

  tree.forEach((node) => {
    seen ? seen.push(node) : flatTree.push(node)
    node.parent = parent
    node.leaf = false

    if (isArray(node.children)) {
      buildTreeRelation(node.children, node, flatTree)
    } else {
      node.leaf = true
    }
  })

  return {
    tree,
    flatTree
  }
}

export function findNodeRelationChain(node) {
  const chain = []
  let parent = node.parent

  while (parent) {
    chain.unshift(parent)
    parent = parent.parent
  }

  return chain
}

export function nextTick() {
  return new Promise((resolve) => Vue.nextTick(resolve))
}

export function doubleRaf() {
  return new Promise((resolve) => {
    window.requestAnimationFrame(() => {
      window.requestAnimationFrame(resolve)
    })
  })
}

