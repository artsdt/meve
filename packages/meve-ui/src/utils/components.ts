import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>;
  unmount(): void;
}

export function mountComponent(component: ComponentOptions<Vue>, container: string, options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}

export function nextTick(): Promise<void> {
  return new Promise((resolve) => Vue.nextTick(resolve))
}
