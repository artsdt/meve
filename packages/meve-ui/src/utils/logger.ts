export function warn(message: string, from?: string) {
  if (process?.env?.NODE_ENV === 'development') {
    console.warn(`Warning: ${message} ${from ? `from ${from}` : ''}`)
  }
}
