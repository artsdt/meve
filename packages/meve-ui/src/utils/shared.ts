export interface CacheInstance<T> {
  cache: T[];

  has(key: T): boolean;

  add(key: T): void;

  remove(key: T): void;

  clear(): void;
}

export const isString = (val: unknown): val is string => typeof val === 'string'

export const toNumber = (val: number | string | boolean | undefined | null): number => {
  if (val == null) return 0

  if (isString(val)) {
    val = parseFloat(val)
    val = Number.isNaN(val) ? 0 : val
    return val
  }

  if (isBoolean(val)) return Number(val)

  return val
}

export const isBoolean = (val: unknown): val is boolean => typeof val === 'boolean'

export const isFunction = (val: unknown): val is Function => typeof val === 'function'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isPlainObject = (val: unknown): val is Record<string, any> =>
  Object.prototype.toString.call(val) === '[object Object]'

export const isObject = (val: unknown): val is object => typeof val === 'object' && val !== null

export const isArray = (val: unknown): val is Array<any> => Array.isArray(val)

export const isEmpty = (val: unknown) =>
  val === undefined || val === null || val === '' || (Array.isArray(val) && !val.length)

export const removeItem = (arr: Array<unknown>, item: unknown) => {
  if (arr.length) {
    const index: number = arr.indexOf(item)
    if (index > -1) {
      return arr.splice(index, 1)
    }
  }
}

export const toggleItem = (arr: Array<unknown>, item: unknown) => {
  arr.includes(item) ? removeItem(arr, item) : arr.push(item)
}

export const throttle = (method: any, mustRunDelay = 200): (() => void) => {
  let timer: number
  let start = 0

  return function loop(this: unknown, ...args) {
    const now = Date.now()
    const elapsed = now - start

    if (!start) {
      start = now
    }

    if (timer) {
      window.clearTimeout(timer)
    }

    if (elapsed >= mustRunDelay) {
      method.apply(this, args)
      start = now
    } else {
      timer = window.setTimeout(() => {
        loop.apply(this, args)
      }, mustRunDelay - elapsed)
    }
  }
}

export const debounce = (method: any, delay = 200) => {
  let timer: number

  return function (this: unknown, ...args: any[]) {
    timer && window.clearTimeout(timer)
    timer = window.setTimeout(() => {
      method.apply(this, args)
    }, delay)
  }
}

export function kebabCase(str: string): string {
  const reg = /([^-])([A-Z])/g

  return str.replace(reg, '$1-$2').replace(reg, '$1-$2').toLowerCase()
}

export const camelize = (s: string): string => s.replace(/-(\w)/g, (_: any, p: string) => p.toUpperCase())

export const bigCamelize = (s: string): string => camelize(s).replace(s.charAt(0), s.charAt(0).toUpperCase())

export const range = (num: number, left: number, right: number) => {
  if (num < left) {
    return left
  }

  if (num > right) {
    return right
  }

  return num
}

export const delay = (time: number) => new Promise((resolve) => setTimeout(resolve, time))

export const inBrowser = () => typeof window !== 'undefined'

export const createCache = <T>(max: number): CacheInstance<T> => {
  const cache: T[] = []

  return {
    cache,

    has(key: T) {
      return this.cache.includes(key)
    },

    add(key: T) {
      if (this.has(key)) {
        return
      }

      this.cache.length === max && cache.shift()
      this.cache.push(key)
    },

    remove(key: T) {
      this.has(key) && removeItem(this.cache, key)
    },

    clear() {
      this.cache.length = 0
    },
  }
}

export const uniq = (arr: Array<any>) => [...new Set(arr)]

export const NOOP = () => {}
