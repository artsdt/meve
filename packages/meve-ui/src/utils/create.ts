import type Vue from 'vue'
import { VNodeData, ComponentOptions, VueConstructor, CreateElement, VNode, RenderContext } from 'vue'
import { PropsDefinition } from 'vue/types/options'
import { bigCamelize, isArray, isFunction, kebabCase } from './shared'
import { SlotsMixin } from './mixins/slots'

export type DefaultProps = Record<string, any>

export type DefaultSlots = {
  default?: ScopedSlot;
}

export type ScopedSlot<Props = any> = (props?: Props) => VNode[] | undefined

export type ScopedSlots = {
  [key: string]: ScopedSlot | undefined;
} & DefaultSlots

export type FunctionComponent<Props = DefaultProps, PropsDefinitions = PropsDefinition<Props>> = {
  (h: CreateElement, props: Props, slots: ScopedSlots, context: RenderContext<Props>): VNode | undefined;
  name?: string;
  props?: PropsDefinitions;
}

export type FunctionComponentPlugin = FunctionComponent & ComponentPlugin

export type TsxBaseProps<Slots> = {
  key: string | number;
  props: any;
  class: any;
  style: string | object[] | object;
  scopedSlots: Slots;
}

export type TsxComponent<Props, Events, Slots> = (props: Partial<Props & Events & TsxBaseProps<Slots>>) => VNode

export interface ComponentPlugin {
  install?: (app: VueConstructor, options?: any) => void;
}

export type ComponentOptionsPlugin = ComponentOptions<Vue> & ComponentPlugin

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function wrapFunctionComponent(
  functionComponent: FunctionComponent
): ComponentOptionsPlugin & { functional?: boolean } {
  return {
    functional: true,
    props: functionComponent.props,
    render: (h: CreateElement, context: RenderContext): any =>
      functionComponent(h, context.props, mergeSlots(context), context),
  }
}

export function install(this: ComponentOptions<Vue>, Vue: VueConstructor) {
  Vue.component(this.name as string, this)
}

export function createComponent<Props = DefaultProps, Events = {}, Slots = {}>(
  name: string,
  sfc: ComponentOptionsPlugin | FunctionComponentPlugin
) {
  if (isFunction(sfc)) {
    sfc = wrapFunctionComponent(sfc as FunctionComponentPlugin)
  } else if ((sfc as ComponentOptionsPlugin).mixins) {
    ;(sfc as ComponentOptionsPlugin).mixins!.push(SlotsMixin)
  } else {
    ;(sfc as ComponentOptionsPlugin).mixins = [SlotsMixin]
  }

  sfc.name = `M${bigCamelize(name)}`
  sfc.install = install

  return sfc as TsxComponent<Props, Events, Slots>
}

export function createNamespace(name: string) {
  return {
    createComponent<Props, Events, Slots>(sfc: ComponentOptionsPlugin | FunctionComponentPlugin) {
      return createComponent<Props, Events, Slots>(name, sfc)
    },
    namespace: (suffix = '') => `m-${kebabCase(name)}${suffix}`,
  }
}

export type Context = RenderContext & { data: VNodeData & Record<string, any> }

export type InheritContext = Partial<VNodeData> & Record<string, any>

const inheritKeys = [
  'ref',
  'key',
  'style',
  'class',
  'attrs',
  'refInFor',
  'nativeOn',
  'directives',
  'staticClass',
  'staticStyle',
]

export function inherit(context: Context): InheritContext {
  return inheritKeys.reduce((ret, key) => {
    if (context.data[key]) {
      ret[key === 'nativeOn' ? 'on' : key] = context.data[key]
    }
    return ret
  }, {} as InheritContext)
}

export function emit(context: RenderContext, eventName: string, ...args: any[]) {
  const listeners = context.listeners[eventName]
  if (isArray(listeners)) {
    listeners.forEach((listener) => listener(...args))
  } else {
    listeners?.(...args)
  }
}
