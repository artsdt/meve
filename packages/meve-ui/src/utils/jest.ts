import Vue from 'vue'
import { Wrapper, config } from '@vue/test-utils'
import { isPlainObject } from './shared'

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function mockOffset({
  offsetWidth,
  offsetHeight,
  clientWidth,
  clientHeight,
  offsetLeft,
  offsetTop,
  scrollWidth,
  scrollHeight,
}: {
  offsetWidth?: number;
  offsetHeight?: number;
  clientWidth?: number;
  clientHeight?: number;
  offsetLeft?: number;
  offsetTop?: number;
  scrollWidth?: number;
  scrollHeight?: number;
} = {}) {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return offsetWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    offsetHeight: {
      get() {
        return offsetHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    clientWidth: {
      get() {
        return clientWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    clientHeight: {
      get() {
        return clientHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    offsetLeft: {
      get() {
        return offsetLeft ?? (parseFloat(window.getComputedStyle(this).marginLeft) || 0)
      },
    },
    offsetTop: {
      get() {
        return offsetTop ?? (parseFloat(window.getComputedStyle(this).marginTop) || 0)
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
    scrollWidth: {
      get() {
        return scrollWidth ?? 0
      },
    },
    scrollHeight: {
      get() {
        return scrollHeight ?? 0
      },
    },
  })

  return {
    mockRestore() {
      Object.defineProperties(HTMLElement.prototype, {
        offsetWidth: {
          get() {
            return parseFloat(window.getComputedStyle(this).width) || 0
          },
        },
        offsetHeight: {
          get() {
            return parseFloat(window.getComputedStyle(this).height) || 0
          },
        },
        clientWidth: {
          get() {
            return parseFloat(window.getComputedStyle(this).width) || 0
          },
        },
        clientHeight: {
          get() {
            return parseFloat(window.getComputedStyle(this).height) || 0
          },
        },
        offsetLeft: {
          get() {
            return parseFloat(window.getComputedStyle(this).marginLeft) || 0
          },
        },
        offsetTop: {
          get() {
            return parseFloat(window.getComputedStyle(this).marginTop) || 0
          },
        },
        offsetParent: {
          get() {
            return this.parentNode ?? {}
          },
        },
        scrollWidth: {
          get() {
            return 0
          },
        },
        scrollHeight: {
          get() {
            return 0
          },
        },
      })
    },
  }
}

export async function triggerDrag(
  el: Wrapper<Vue> | HTMLElement | Window,
  x = 0,
  y = 0,
  triggerTouchend = true,
  triggerDocument = false
) {
  const delegationEl = triggerDocument ? document : el

  await trigger(el, 'touchstart', 0, 0)
  await trigger(delegationEl, 'touchmove', x / 4, y / 4)
  await trigger(delegationEl, 'touchmove', x / 3, y / 3)
  await trigger(delegationEl, 'touchmove', x / 2, y / 2)
  await trigger(delegationEl, 'touchmove', x, y)

  if (triggerTouchend) {
    await trigger(delegationEl, 'touchend', x, y)
  }
}

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  deltaX = 0,
  deltaY = 0,
  offsetX = 0,
  offsetY = 0,
  args = {}
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    deltaX,
    deltaY,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
    ...args,
  })

  el.dispatchEvent(event)

  return Vue.nextTick()
}

export function mockStubs() {
  const originStubs = config.stubs

  config.stubs = {}

  return {
    mockRestore() {
      config.stubs = originStubs
    },
  }
}

export function mockDoubleRaf() {
  const originMethod = window.requestAnimationFrame

  Object.assign(window, {
    requestAnimationFrame(fn: any) {
      setTimeout(fn, 16)
    },
  })

  return {
    mockRestore() {
      window.requestAnimationFrame = originMethod
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
