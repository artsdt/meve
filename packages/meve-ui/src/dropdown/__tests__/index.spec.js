import Dropdown from '..'
import { mount } from '@vue/test-utils'
import { delay, mockOffset, trigger } from '../../utils/jest'

mockOffset()

test('test dropdown open', async () => {
  const wrapper = mount(Dropdown, {
    propsData: {
      options: [
        { key: 1, label: 'hello' },
        { key: 2, label: 'world' },
      ],
    },
    slots: {
      default: '<button class="trigger">hello</button>',
    },
    attachTo: document.body,
  })

  await wrapper.find('.trigger').trigger('click')
  await delay(100)

  expect(document.body.innerHTML.includes('display: none')).toBe(false)

  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test dropdown size', async () => {
  const wrapper = mount(Dropdown, {
    propsData: {
      options: [
        { key: 1, label: 'hello', icon: 'error' },
        { key: 2, label: 'world', icon: 'error' },
      ],
      size: 'small',
    },
    slots: {
      default: '<button class="trigger">hello</button>',
    },
    attachTo: document.body,
  })

  expect(document.body.innerHTML).toMatchSnapshot()
  expect(document.body.innerHTML.includes('m-dropdown--small')).toBe(true)

  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test dropdown select', async () => {
  const wrapper = mount(Dropdown, {
    propsData: {
      options: [
        { key: 1, label: 'hello' },
        { key: 2, label: 'world' },
      ],
    },
    slots: {
      default: '<button class="trigger">hello</button>',
    },
    attachTo: document.body,
  })

  await wrapper.find('.trigger').trigger('click')
  await delay(100)
  expect(wrapper.emitted().open.length).toBe(1)
  await trigger(document.querySelector('.m-dropdown__cell'), 'click')
  expect(wrapper.emitted().select.length).toBe(1)
  expect(wrapper.emitted().select[0][0].key).toBe(1)
  await delay(300)
  expect(wrapper.emitted().close.length).toBe(1)
  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test dropdown select disabled', async () => {
  const wrapper = mount(Dropdown, {
    propsData: {
      options: [
        { key: 1, label: 'hello', disabled: true },
        { key: 2, label: 'world' },
      ],
    },
    slots: {
      default: '<button class="trigger">hello</button>',
    },
    attachTo: document.body,
  })

  await wrapper.find('.trigger').trigger('click')
  await delay(100)
  await trigger(document.querySelector('.m-dropdown__cell'), 'click')
  expect(wrapper.emitted().select).toBeFalsy()
  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test dropdown custom render', async () => {
  const wrapper = mount(Dropdown, {
    propsData: {
      options: [
        {
          key: 1,
          label: (h) => h('span', 'custom hello'),
          icon: (h) => h('span', 'custom icon'),
        },
      ],
    },
    slots: {
      default: '<button class="trigger">hello</button>',
    },
    attachTo: document.body,
  })

  expect(document.body.innerHTML).toMatchSnapshot()
  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test dropdown children', async () => {
  const wrapper = mount(Dropdown, {
    propsData: {
      options: [
        {
          key: 1,
          label: 'parent',
          children: [
            {
              key: 2,
              label: 'child',
            },
          ],
        },
      ],
    },
    slots: {
      default: '<button class="trigger">hello</button>',
    },
    attachTo: document.body,
  })

  console.log(document.body.innerHTML)

  const childNode = document.querySelectorAll('.m-dropdown__cell')[0]
  const parentNode = document.querySelectorAll('.m-dropdown__cell')[1]

  await wrapper.find('.trigger').trigger('click')
  await delay(300)
  await trigger(parentNode, 'click')
  await delay(300)
  expect(document.body.innerHTML).toMatchSnapshot()

  await trigger(childNode, 'click')
  expect(wrapper.emitted().select.length).toBe(1)
  expect(wrapper.emitted().select[0][0].key).toBe(2)
  wrapper.destroy()
  document.body.innerHTML = ''
})
