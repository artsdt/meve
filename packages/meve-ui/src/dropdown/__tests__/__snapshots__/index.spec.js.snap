// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test dropdown children 1`] = `"<div class=\\"m-dropdown\\"><div class=\\"m-popover\\"><button class=\\"trigger\\">hello</button><div class=\\"m-teleport\\"></div></div></div><div><div class=\\"m-popover__container\\" style=\\"transform: scale(1); transform-origin: right top; z-index: 2012; top: 0px; left: 4px;\\"><div class=\\"m-dropdown__wrapper\\"><div class=\\"m-dropdown__cell m-dropdown--normal\\"><div class=\\"m-dropdown__label\\">child</div></div></div></div></div><div><div class=\\"m-popover__container\\" style=\\"transform: scale(1); transform-origin: bottom; z-index: 2011; top: 4px; left: 0px;\\"><div class=\\"m-dropdown__wrapper\\"><div class=\\"m-popover\\"><div class=\\"m-dropdown__cell m-dropdown--normal\\"><div class=\\"m-dropdown__label\\">parent</div><i dropdown-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-arrow-right m-dropdown__arrow\\"></i></div><div class=\\"m-teleport\\"></div></div></div></div></div>"`;

exports[`test dropdown custom render 1`] = `"<div class=\\"m-dropdown\\"><div class=\\"m-popover\\"><button class=\\"trigger\\">hello</button><div class=\\"m-teleport\\"></div></div></div><div><div class=\\"m-popover__container\\" style=\\"transform: scale(0.8); transform-origin: top; opacity: 0; z-index: 2007; display: none;\\"><div class=\\"m-dropdown__wrapper\\"><div class=\\"m-dropdown__cell m-dropdown--normal\\"><span>custom icon</span><span>custom hello</span></div></div></div></div>"`;

exports[`test dropdown size 1`] = `"<div class=\\"m-dropdown\\"><div class=\\"m-popover\\"><button class=\\"trigger\\">hello</button><div class=\\"m-teleport\\"></div></div></div><div><div class=\\"m-popover__container\\" style=\\"transform: scale(0.8); transform-origin: top; opacity: 0; z-index: 2002; display: none;\\"><div class=\\"m-dropdown__wrapper\\"><div class=\\"m-dropdown__cell m-dropdown--small\\"><i dropdown-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-error m-dropdown__icon\\"></i><div class=\\"m-dropdown__label\\">hello</div></div><div class=\\"m-dropdown__cell m-dropdown--small\\"><i dropdown-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-error m-dropdown__icon\\"></i><div class=\\"m-dropdown__label\\">world</div></div></div></div></div>"`;
